### PR TITLE
feat(STONEINTG-725): retry for recoverable error when creating relase

### PR DESCRIPTION
**Pull request by hongweiliu17**
when creating release for snapshot
* retry for recoverable error

Signed-off-by: Hongwei Liu <hongliu@redhat.com>

## Maintainers will complete the following section

- [ ] Commit messages are descriptive enough ([hints](https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/))
- [ ] Code coverage from testing does not decrease and new code is covered ([check the PR coverage on codecov](https://app.codecov.io/gh/redhat-appstudio/integration-service/pulls))
- [ ] [Controllers diagrams](https://github.com/redhat-appstudio/integration-service/tree/main/docs) are updated when PR changes controllers code  (if applicable)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -16,7 +16,7 @@ testbin/*
 cover.html
 
 # folder generated by go mod vendor
-vendor
```
Comment (MartinBasti) [inline .gitignore:19]: Why this change?

File: controllers/snapshot/snapshot_adapter.go [hunk 2]
```diff
@@ -423,7 +429,16 @@ func (a *Adapter) EnsureAllReleasesExist() (controller.OperationResult, error) {
...
 		a.logger.LogAuditEvent("Snapshot integration status marked as Invalid. Failed to create new Releases",
 			a.snapshot, h.LogActionUpdate)
-		return controller.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.snapshot, patch))
+		er := a.client.Status().Patch(a.context, a.snapshot, patch)
+		if er != nil {
+			a.logger.Error(er, "Failed to mark snapshot integration status as invalid",
+				"snapshot.Name", a.snapshot.Name)
+			return controller.RequeueWithError(er)
+		}
+		if clienterrors.IsForbidden(err) {
```
Comment (MartinBasti) [inline controllers/snapshot/snapshot_adapter.go:438]: Why this specific error is considered to be non-recoverable?

File: controllers/snapshot/snapshot_adapter.go [hunk 3]
```diff
@@ -413,7 +413,13 @@ func (a *Adapter) EnsureAllReleasesExist() (controller.OperationResult, error) {
 		gitops.SetSnapshotIntegrationStatusAsError(a.snapshot, "Failed to get all ReleasePlans: "+err.Error())
 		a.logger.LogAuditEvent("Snapshot integration status marked as Invalid. Failed to get all ReleasePlans",
 			a.snapshot, h.LogActionUpdate)
-		return controller.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.snapshot, patch))
+		er := a.client.Status().Patch(a.context, a.snapshot, patch)
+		if er != nil {
+			a.logger.Error(er, "Failed to mark snapshot integration status as invalid",
+				"snapshot.Name", a.snapshot.Name)
+			return controller.RequeueWithError(er)
```
Comment (MartinBasti) [inline controllers/snapshot/snapshot_adapter.go:420]: This hides why we even get to this error, we should expose both errors, we can use Join function https://pkg.go.dev/errors#Join

File: .gitignore (hunk 1 above)
Comment (hongweiliu17) [reply .gitignore:19]: I removed it when I read source code. I will add it back and rebase on main branch together after https://github.com/redhat-appstudio/integration-service/pull/482 is merged. Thanks!

File: controllers/snapshot/snapshot_adapter.go (hunk 2 above)
Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go:438]: I think this forbidden error need some operation to make it work again, e.g., the quota limitation is reached.

File: controllers/snapshot/snapshot_adapter.go (hunk 3 above)
Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go:420]: Expose? Do you mean expose `er` and `err` here together? The `err` is logged out above and will retry `controller.RequeueWithError(err)` below

Comment (MartinBasti) [reply controllers/snapshot/snapshot_adapter.go:420]: Yes, exposing errors together

File: controllers/snapshot/snapshot_adapter.go (hunk 2 above)
Comment (MartinBasti) [reply controllers/snapshot/snapshot_adapter.go:438]: I think that quota limit should be kept in reconciliation, so once old objects are emptied, it will recover with processing.

Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go:438]: But users might notice the issue several days later, at that time we might have stopped the reconciliation after failing several times.

Comment (MartinBasti) [reply controllers/snapshot/snapshot_adapter.go:438]: but if they recover, they would need to wait one day

Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go:438]: So let's remove this stop and see users notice this limitation generally then handle it in [STONEINTG-729](https://issues.redhat.com/browse/STONEINTG-729)

**Review by MartinBasti — APPROVED** [review]

**Comment by hongweiliu17** [comment]
I trigger e2e test in [e2e-tests PR](https://github.com/redhat-appstudio/e2e-tests/pull/980)  and integration-service-e2e in this PR at the same time, but the e2e in [e2e-tests PR](https://github.com/redhat-appstudio/e2e-tests/pull/980) passes, the test in this PR fails

**Comment by kasemAlem** [comment]
> I trigger e2e test in [e2e-tests PR](https://github.com/redhat-appstudio/e2e-tests/pull/980) and integration-service-e2e in this PR at the same time, but the e2e in [e2e-tests PR](https://github.com/redhat-appstudio/e2e-tests/pull/980) passes, the test in this PR fails

**Comment by hongweiliu17** [comment]
Look at the recent three cycle of integration-service-e2e test for this PR, one suceed, two failed.
image
It looks like timeout, we may try to improve it.

Red Hat App Studio E2E tests: [It] [integration-service-suite Namespace-backed Environment (NBE) E2E tests] with happy path for Namespace-backed environments when Integration Test PipelineRun completes successfully should lead to Snapshot CR being marked as passed [integration-service, HACBS, namespace-backed-envs] expand_less 	9m3s
{Timed out after 180.001s.
Timed out waiting for Snapshot to be marked as succeeded nbe-e2e-zbja-tenant/integ-app-nmjn-9wt9q
Expected
    <bool>: false
to be true failed [FAILED] Timed out after 180.001s.
Timed out waiting for Snapshot to be marked as succeeded nbe-e2e-zbja-tenant/integ-app-nmjn-9wt9q
Expected
    <bool>: false
to be true
In [It] at: /tmp/tmp.pKqnXrFzKX/tests/integration-service/namespace-backed-environments.go:246 @ 12/25/23 09:20:38.536

There were additional failures detected after the initial failure. These are visible in the timeline
}
[open stderropen_in_new](https://prow.ci.openshift.org/spyglass/lens/junit/iframe?req=%7B%22artifacts%22%3A%5B%22artifacts%2Fintegration-service-e2e%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_install_status.xml%22%2C%22artifacts%2Fintegration-service-e2e%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_symptoms.xml%22%2C%22artifacts%2Fintegration-service-e2e%2Fredhat-appstudio-report%2Fartifacts%2Fjunit.xml%22%2C%22artifacts%2Fjunit_operator.xml%22%5D%2C%22index%22%3A2%2C%22src%22%3A%22gs%2Forigin-ci-test%2Fpr-logs%2Fpull%2Fredhat-appstudio_integration-service%2F481%2Fpull-ci-redhat-appstudio-integration-service-main-integration-service-e2e%2F1739204438700068864%22%7D&topURL=https%3A//prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/redhat-appstudio_integration-service/481/pull-ci-redhat-appstudio-integration-service-main-integration-service-e2e/1739204438700068864&lensIndex=2#)
: redhat-appstudio-e2e expand_less 	0s
{redhat-appstudio-e2e has failed  }
[open stderropen_in_new](https://prow.ci.openshift.org/spyglass/lens/junit/iframe?req=%7B%22artifacts%22%3A%5B%22artifacts%2Fintegration-service-e2e%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_install_status.xml%22%2C%22artifacts%2Fintegration-service-e2e%2Fgather-extra%2Fartifacts%2Fjunit%2Fjunit_symptoms.xml%22%2C%22artifacts%2Fintegration-service-e2e%2Fredhat-appstudio-report%2Fartifacts%2Fjunit.xml%22%2C%22artifacts%2Fjunit_operator.xml%22%5D%2C%22index%22%3A2%2C%22src%22%3A%22gs%2Forigin-ci-test%2Fpr-logs%2Fpull%2Fredhat-appstudio_integration-service%2F481%2Fpull-ci-redhat-appstudio-integration-service-main-integration-service-e2e%2F1739204438700068864%22%7D&topURL=https%3A//prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/redhat-appstudio_integration-service/481/pull-ci-redhat-appstudio-integration-service-main-integration-service-e2e/1739204438700068864&lensIndex=2#)
: Run multi-stage test integration-service-e2e expand_less 	52m21s
{  "integration-service-e2e" test steps failed: "integration-service-e2e" pod "integration-service-e2e-redhat-appstudio-e2e" failed: could not watch pod: the pod ci-op-b7d50mnn/integration-service-e2e-redhat-appstudio-e2e failed after 38m2s (failed containers: test): ContainerFailed one or more containers exited

Container test exited with code 2, reason Error
---
ecs in 1264.017 seconds�[0m
�[38;5;9m�[1mFAIL!�[0m -- �[38;5;10m�[1m60 Passed�[0m | �[38;5;9m�[1m1 Failed�[0m | �[38;5;11m�[1m16 Pending�[0m | �[38;5;14m�[1m548 Skipped�[0m


Ginkgo ran 1 suite in 22m23.638934774s

**Comment by hongweiliu17** [comment]
https://github.com/redhat-appstudio/e2e-tests/pull/981 is filed to try fix the timeout in NPE case.

**Review by kasemAlem — APPROVED** [review]

**Review by dirgim — APPROVED** [review]
LGTM
